store/inmemory_backing_store.go

```diff
@@ -0,0 +1,143 @@
+package store
+
+import (
+	"errors"
+	"github.com/google/uuid"
+	"reflect"
+	"strings"
+)
+
+type BackingStoreSubscriber func(key string, oldVal interface{}, newVal interface{})
+
+type inMemoryBackingStore struct {
+	returnOnlyChangedValues bool
+	initializationCompleted bool
+	store                   map[string]interface{}
+	subscribers             map[string]BackingStoreSubscriber
+	changedValues           map[string]bool
+}
+
+// NewInMemoryBackingStore returns a new instance of an in memory backing store
+func NewInMemoryBackingStore() BackingStore {
+	return &inMemoryBackingStore{
+		returnOnlyChangedValues: false,
+		initializationCompleted: true,
+		store:                   make(map[string]interface{}),
+		subscribers:             make(map[string]BackingStoreSubscriber),
+		changedValues:           make(map[string]bool),
+	}
+}
+
+func (i *inMemoryBackingStore) Get(key string) (interface{}, error) {
+	key = strings.TrimSpace(key)
+	if key == "" {
+		return nil, errors.New("key cannot be an empty string")
+	}
+
+	objectVal := i.store[key]
+
+	if (i.GetReturnOnlyChangedValues() && i.changedValues[key]) || !i.GetReturnOnlyChangedValues() {
+		return objectVal, nil
+	} else {
+		return nil, nil
+	}
+}
+
+func (i *inMemoryBackingStore) Set(key string, value interface{}) error {
+	key = strings.TrimSpace(key)
+	if key == "" {
+		return errors.New("key cannot be an empty string")
+	}
+
+	current := i.store[key]
+
+	// check if objects values have changed
+	if current == nil || !reflect.DeepEqual(current, value) {
```

Comment:
can you expand on the rationale for the deep comparison here?

Reply:
In such cases, I do not have to loop over collections and check for individual object differences, using deep equal makes it simpler to compare collections or collections nested in object with collection properties

Reply:
Thank for the additional context here. I do think this is unique to Go at this point. All the other implementations assume that if somebody wants to update a collection, they'll update the reference to the collection.
The reasoning behind this is that deep compares are compute intensive, and we might come to it only if the feedback on dirty tracking asks for it. Thoughts?

Reply:
I do agree, was on the fence with just how much computation reflection takes in Go, we can dump the reflection comparison. However, reading up on append had me a bit worried since it looks like adding elements to an existing slice might or might not return a new slice. 

> The append built-in function appends elements to the end of a slice. If it has sufficient capacity, the destination is resliced to accommodate the new elements. If it does not, a new underlying array will be allocated. Append returns the updated slice.
> source : https://pkg.go.dev/builtin#append

Given the rest of the languages make the same assumption I think we can do the same

Reply:
right, but they'll have to use the setter since they can't make the assumption the slice will remain unchanged. And as they do that, our tracking code will be triggered.